src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -624,17 +625,17 @@ internal void CancelPendingFlush()
             TrySchedule(_writerScheduler, completionData);
         }
 
-        internal void OnReaderCompleted(Action<Exception, object> callback, object state)
+        internal void OnReaderCompleted(Action<Exception?, object?> callback, object? state)
         {
             if (callback is null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.callback);
             }
 
-            PipeCompletionCallbacks completionCallbacks;
+            PipeCompletionCallbacks? completionCallbacks;
             lock (_sync)
             {
-                completionCallbacks = _readerCompletion.AddCallback(callback, state);
+                completionCallbacks = _readerCompletion.AddCallback(callback!, state);
```

Comment:
Another case where the `!` shouldn't be necessary.  Once you add the `[DoesNotReturn]` to the ThrowHelper methods, you can remove this.